### PR TITLE
First iteration of adding models for renderers/views; pull out events

### PR DESCRIPTION
**Pull request by jleft**

### CODE REVIEW

File: src/assets/js/chart/primary.js [hunk 1]
```diff
@@ -154,15 +154,26 @@
...
+
+        function changeYValueAccessor(yValueAccessor) {
+            currentYValueAccessor = yValueAccessor.option;
+        }
+
         function primary(selection) {
             var model = selection.datum();
+            changeSeries(model.series);
+            changeYValueAccessor(model.yValueAccessor);
+            currentIndicators = model.indicators;
```
Comment (djmiley) [inline src/assets/js/chart/primary.js]: Why extract `changeSeries` as a function but not this?

File: src/assets/js/app.js [hunk 2]
```diff
@@ -62,68 +60,77 @@
...
-            model.viewDomain = [domain[0], domain[1]];
-            model.trackingLatest = sc.util.domain.trackingLatestData(model.viewDomain, model.data);
+        function onViewChange(domain) {
+            primaryChartModel.viewDomain = [domain[0], domain[1]];
+            secondaryChartModel.viewDomain = [domain[0], domain[1]];
+            xAxisModel.viewDomain = [domain[0], domain[1]];
+            navModel.viewDomain = [domain[0], domain[1]];
+            primaryChartModel.trackingLatest = sc.util.domain.trackingLatestData(
+                primaryChartModel.viewDomain,
+                primaryChartModel.data);
```
Comment (djmiley) [inline src/assets/js/app.js]: I think all the other charts (except `sc.chart.xAxis`) need to know about `trackingLatest` in order to stick to the latest data when zooming in.

File: src/assets/js/app.js [hunk 3]
```diff
@@ -62,68 +60,77 @@
...
                     if (socketEvent.type === 'error' ||
                         (socketEvent.type === 'close' && socketEvent.code !== 1000)) {
                         console.log('Error loading data from coinbase websocket: ' +
                         socketEvent.type + ' ' + socketEvent.code);
                     } else if (socketEvent.type === 'message') {
                         updateModelData(data);
-                        if (model.trackingLatest) {
-                            var newDomain = sc.util.domain.moveToLatest(model.viewDomain, model.data);
-                            onViewChanged(newDomain);
+                        if (primaryChartModel.trackingLatest) {
```
Comment (djmiley) [inline src/assets/js/app.js]: It's probably right that `primaryChartModel` would be our default frame of reference out of all the charts. Is this the intention here?

File: src/assets/js/app.js [hunk 4]
```diff
@@ -62,68 +60,77 @@
...
-        function initialiseChartEventHandlers() {
-            primaryChart.on('crosshairChange', onCrosshairChanged);
-            primaryChart.on('viewChange', onViewChanged);
-            nav.on('viewChange', onViewChanged);
-        }
-
-        function resetToLive() {
-            var data = model.data;
+        function resetToLatest() {
+            var data = primaryChartModel.data;
```
Comment (djmiley) [inline src/assets/js/app.js]: See below.

**Comment by djmiley** [comment]
Like the refactoring, looks really good. :) Strikes me that we may want a basic chart model (with `data` and `viewDomain` on, maybe an  basic interactive chart model with `trackingLatest` on in addition) that extends to more specialised implementations.

File: src/assets/js/app.js [hunk 5]
```diff
@@ -131,40 +138,38 @@
...
-                .on('primaryChartYValueAccessorChange', function(yValueAccessor) {
-                    primaryChart.changeYValueAccessor(yValueAccessor);
+                .on(sc.event.primaryChartYValueAccessorChange, function(yValueAccessor) {
+                    primaryChartModel.yValueAccessor = yValueAccessor;
                     render();
                 })
-                .on('primaryChartIndicatorChange', function(toggledIndicator) {
-                    primaryChart.toggleIndicator(toggledIndicator);
+                .on(sc.event.primaryChartIndicatorChange, function(toggledIndicator) {
+                    primaryChartModel.toggleIndicator(toggledIndicator);
```
Comment (djmiley) [inline src/assets/js/app.js]: Could pull the `toggleIndicator` code out of `primaryChartModel` (near-exact duplication of lines 168-174), then set the new array of `indicators` on the model.

Comment (djmiley) [inline src/assets/js/app.js]: Doesn't feel `toggleIndicator` belongs on the `primaryChartModel`

Comment (tlsim) [inline src/assets/js/app.js]: @jleft do you see these more as view models in the MVVM sense?

File: src/assets/js/chart/primary.js (hunk 1 above)
Comment (jleft) [inline src/assets/js/chart/primary.js]: Not particular reason, I'll update to be consistent.

File: src/assets/js/app.js (hunk 2 above)
Comment (jleft) [inline src/assets/js/app.js]: Good catch :+1:

File: src/assets/js/app.js (hunk 5 above)
Comment (jleft) [inline src/assets/js/app.js]: > @jleft do you see these more as view models in the MVVM sense?

Yeah, I'm leaning that way at the moment.

Comment (djmiley) [inline src/assets/js/app.js]: As discussed in kitchen, I have reconsidered my opinion on this, by considering implementation for indicators consisting of overlays and secondary charts.

**Comment by jleft** [comment]
> Like the refactoring, looks really good. :) Strikes me that we may want a basic chart model (with data and viewDomain on, maybe an basic interactive chart model with trackingLatest on in addition) that extends to more specialised implementations.

Thanks @djmiley :). Yeah, I've had similar thoughts - I can create an issue to track the idea?

**Comment by jleft** [comment]
Incorporated in #297

**Pull request closed without merging**